Respecting categorized ordering and persisting this upon drag and drop when categories are enabled

payten opened the pull request:
I've reintroduced the order XML but slightly tweaked it to take into account the assignment's category.  Now when an the page is loaded, we grab this XML and poke the corresponding order onto the header cell so the Javascript can use it when reordering the cells upon enabling categories.

When a drag and drop is performed when categories are enabled, we persist this order via a new endpoint that updates the XML values and stores it.

There's a bit of Java wrangling in this one, so @steveswinsburg it could use a good review!  Thanks :cake:

payten committed:
Introduce categorized assignment order and store this on the gradebook site's properties

payten committed:
Expose the categorized sort order in the DOM so the javascript can grab it and do things

payten committed:
Update Javascript to order columns based on the element attribute data-catagorized-order and also to hit the new API endpoint to update the categorized sort order (when categories are enabled and a drag and drop is performed)

steveswinsburg commented:
I am looking at this just want to give it a good review so should hopefully have it in tomorrow.

steveswinsburg commented on tool/src/java/org/sakaiproject/gradebookng/tool/pages/GradebookPage.java:
```diff
@@ -146,6 +148,10 @@ public String getCssClass() {
+                String category = assignment.getCategoryName();
+                int order = categorizedAssignmentOrder.get(category).indexOf(assignment.getId());
+                panel.add(new AttributeModifier("data-category", category));
+                panel.add(new AttributeModifier("data-catagorized-order", order));
```
Noticed this typo, catagorized

payten replied:
Oops.. sometimes I lapse into American spelling.

steveswinsburg commented on tool/src/webapp/scripts/gradebook-grades.js:
```diff
@@ -1162,6 +1177,11 @@ GradebookHeaderCell.prototype.hide = function() {
 
 
+GradebookHeaderCell.prototype.getCategorizedOrder = function() {
+  return this.$cell.find("[data-catagorized-order]").data("catagorized-order");
```
Same typo, catagorized

steveswinsburg replied:
Nah I meant the catAgor... ;)

payten replied:
Touché!

steveswinsburg commented:
What do you think the performance impact is going to be with this XML storage of the order? I am wondering if this is another candidate for the main gradebook service. There is an order in there that we might be able to leverage in the same way I did the assignment order. 

This would give us some performance wins. I'm just concerned about adding the XML back in for this since this is a high intensity operation for the loading. I think we are going to need it for the other config data that we have, still in two minds about having it for this.

Maybe we add the logging/timing statements first so we can see if I am worrying about nothing.

Is the category order meant to be site wide or per user? If if an instructor sets it, should everyone need the new order? That would be the driver. If the answer was 'no it should be for the user only', then we need the XML storage.

payten commented:
You're totally right.  It'd be great to keep any XML parsing off the critical path of the page load (I've seen what it does to other tools!).  

I have a feeling this is meant to be a per-user feature.  @jeffpasch - is that correct?

If this can't be moved to the database easily, could we look into making the frontend Javascript load the category information/ordering via an Ajax request?  The category button would be disabled (or show a spinner?) and the categorised drag and drop would only be initialised/enabled until the category data was loaded.

jeffpasch commented:
Yup, the ordering should just apply to a given user, not the site.

steveswinsburg commented:
Hmm yeah an Ajax request afterwards might be a goer, but I'll get some timing data in first so we can see if it makes a diff anyway. I'll then merge it into this and get you to do the same so we can run some tests and see if there really is any performance issue to worry about with the XML serialisation etc.

payten committed:
Fix spelling mistakes

payten commented:
Sure thing @steveswinsburg.  Keep me posted and I'll run some comparison tests when ready to roll.

payten commented:
Hm.. if all the sorting is user-based rather than site-based, @steveswinsburg will we need to key this XML on the user ID?  Correct me if I'm wrong, but are both orders currently site-scoped?

payten commented:
Hi @steveswinsburg, just wondering if you think I should refactor this to be user-centered?  Is it ok to continue to store on the tool properties but key on user? Or perhaps move to the user properties?

payten commented:
Rebased for a happy commit.  Both orders are still system wide at the moment.  Perhaps we can discuss a mechanism for storing this per-user next call. 

I've also been thinking that we'll need to handle when assignments are deleted or their categories are changed.  Any ideas?  Seems like we'd need to rebuild the XML for all users at that point.. or upon read perhaps.

steveswinsburg commented on tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java:
```diff
@@ -552,8 +552,169 @@ public void updateAssignmentOrder(String siteId, long assignmentId, int order) {
+  public void updateCategorizedAssignmentOrder(String siteId, long assignmentId, int order) throws JAXBException, IdUnusedException, PermissionException {
+    Gradebook gradebook = (Gradebook)gradebookService.getGradebook(siteId);
+    Assignment assignmentToMove = gradebookService.getAssignment(gradebook.getUid(), assignmentId);
+    String category = assignmentToMove.getCategoryName();
```
This may throw a NPE if null, null check is afterwards, probably needs to be before.

steveswinsburg commented:
Sorry for the silence on this. Lets leave it site scoped for now and see how it comes out in testing. User props we may stick on the user props table instead of the site props table.

For removing assignments, the current order should still be ok wouldn't it? The order is only retrieved for the given assignment so the deleted ones would be in the list but ignored? Then the next reorder would fix it up. Have you tested what happens in this scenario?

payten committed:
Add null checks when retrieving the site and gradebook in the get, update and store methods for categorized ordering.

Also fix NPE when new assignment with category and neither are present in the categorized ordering XML.  Assignments not currently in the XML have an index of -1 and are ordered to the right of those assignments with a valid index.

payten commented:
Thanks @steveswinsburg.  I've added some NPE handling around `site` and `gradebook`. 

After some testing with adding, removing and changing categories I've updated the category ordering code to handle when an assignment doesn't have a categorized order in the XML or the category isn't present in the XML.
